### PR TITLE
Exclude PluginManager cache from MaxMemoryPreload monitoring

### LINKED ISSUE #1074: Exclude PluginManager cache from MaxMemoryPreload monitoring

### LINKED ISSUE #46359: MaxMemoryPreload monitoring shows noticeable increase when large number of packages are rebuilt
Observed in https://github.com/cms-sw/cmssw/pull/46200#issuecomment-2394356850.

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

As described in https://github.com/cms-sw/cmssw/issues/46359, developer areas with many checked out packages result in larger "max memory" reported by MaxMemoryPreload AllocMonitor. This additional memory is used by PluginManager cache.

As a simple (albeit a bit hacky) workaround, this PR adds hooks to the MaxMemoryPreload that allow its data collection to be paused, and uses those hooks in PluginManager to paused the data collection during the cache file reading. 

Resolves https://github.com/cms-sw/cmssw/issues/46359
Resolves https://github.com/cms-sw/framework-team/issues/1074

#### PR validation:

Tested the impact with two developer areas of CMSSW_14_2_0_pre2: one with many packages checked out, and one with only two packages (`FWCore/PluginManager` and `PerfTools/MaxMemoryPreload`). The remaining difference in "max memory" was about 1 MB, or 0.05 %.

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@Dr15Jones please review

File: FWCore/PluginManager/src/PluginManager.cc [hunk 1]
```diff
@@ -47,7 +49,9 @@ namespace edmplugin {
         throw cms::Exception("PluginMangerCacheProblem")
             << "Unable to open the cache file '" << cacheFile.string() << "'. Please check permissions on file";
       }
+      pauseMaxMemoryPreload();
```
Comment (Dr15Jones) [inline FWCore/PluginManager/src/PluginManager.cc:52]: Could we do this with some sort of sentinel? That way if an exception happens it gets turned back on.

File: FWCore/PluginManager/src/pauseMaxMemoryPreload.h [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#ifndef FWCore_PluginManager_src_pauseMaxMemoryPreload_h
+#define FWCore_PluginManager_src_pauseMaxMemoryPreload_h
+
+// By default do nothing, but add "hooks" that MaxMemoryPreload can
+// override with LD_PRELOAD
+void pauseMaxMemoryPreload();
+void unpauseMaxMemoryPreload();
```
Comment (Dr15Jones) [inline FWCore/PluginManager/src/pauseMaxMemoryPreload.h:7]: How about designing a sentinel class here to auto unpause?

File: FWCore/PluginManager/src/PluginManager.cc (hunk 1 above)
Comment (makortel) [reply FWCore/PluginManager/src/PluginManager.cc]: Done

File: FWCore/PluginManager/src/pauseMaxMemoryPreload.h (hunk 2 above)
Comment (makortel) [reply FWCore/PluginManager/src/pauseMaxMemoryPreload.h]: Done

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
Comparison failures are related to https://github.com/cms-sw/cmssw/issues/46416

**Comment by makortel** [comment]
@cmsbuild, please test

Removed spurious `;`

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
Comparison differences are related to https://github.com/cms-sw/cmssw/issues/46416

**Comment by makortel** [comment]
To document the level of reduction in max memory
image

File: FWCore/PluginManager/src/PauseMaxMemoryPreloadSentry.cc [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#include "PauseMaxMemoryPreloadSentry.h"
+
+// By default do nothing, but add "hooks" that MaxMemoryPreload can
+// override with LD_PRELOAD
+void pauseMaxMemoryPreload() {}
+void unpauseMaxMemoryPreload() {}
```
Comment (makortel) [inline FWCore/PluginManager/src/PauseMaxMemoryPreloadSentry.cc:6]: Maybe better to make these `extern "C"`

Comment (makortel) [reply FWCore/PluginManager/src/PauseMaxMemoryPreloadSentry.cc:6]: Actually no, it shouldn't matter here, where we and our C++ code is responsible of both ends.

**Comment by makortel** [comment]
+core

**Comment by makortel** [comment]
ignore tests-rejected with external-failure

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**